Draft PR: Ecash (#1093)
BitcoinDesign/Guide: guide/how-it-works/ecash/fedimint.md

#### Pros:

* **Security** - No single guardian can control the mint, reducing the risk of theft or mismanagement.

* **High resilience** - The mint remains operational even if one or more guardians are temporarily offline, ensuring continuous access.

#### Cons:

* **Complex setup** - Coordinating multiple guardians adds steps and cost.

### Federated single guardian mints

{% include picture.html
   image = "/assets/images/guide/how-it-works/ecash/single-mint@1x.jpg"
   retina = "/assets/images/guide/how-it-works/ecash/single-mint@2x.jpg"
   mobile = "/assets/images/guide/how-it-works/ecash/single-mintmobile@2x.jpg"
   mobileRetina = "assets/images/guide/how-it-works/ecash/single-mint-mobile@2x.jpg"
   alt-text = "An illustration of how users, ecash, and a single guardian mint interact."
   caption = "How a guardian interacts with users of a single guardian mint."
   width = 800
   height = 456
   layout = "full-width"
%}

This model offers a compromise between security and simplicity. A single guardian manages the mint but runs multiple servers, providing some level of resilience without the need for multiple guardians.

#### Pros:

> * **Enhanced Resilience** - Even if some servers fail, the mint remains operational, offering a good balance of security and simplicity.

Author: What do you think about this: 

_By using servers from multiple providers, the mint remains operational even during scheduled downtime or attempts of censorship, ensuring both reliability and security._

> * **Enhanced Resilience** - Even if some servers fail, the mint remains operational, offering a good balance of security and simplicity.

Reviewer: Sounds good :)

#### Cons:

* **Security Trade-off** - With only one guardian, there’s still a single point of failure, making it less secure than a fully federated setup.

### Unfederated Mints (Solo Mints)

{% include picture.html
   image = "/assets/images/guide/how-it-works/ecash/solomint@1x.jpg"
   retina = "/assets/images/guide/how-it-works/ecash/solomint@2x.jpg"
   mobile = "/assets/images/guide/how-it-works/ecash/solomint-mobile@2x.jpg"
   mobileRetina = "assets/images/guide/how-it-works/ecash/solomint-mobile@2x.jpg"
   alt-text = "An illustration of how users, ecash, and a solo mint interact."
   caption = "How a guardian interacts with users of a solo mint."
   width = 800
   height = 456
   layout = "full-width"
%}

Unfederated Mints are the simplest and most cost-effective option. Here, a single guardian manages the entire mint, making it easy to set up but with significant trade-offs in terms of security and resilience.

#### Pros:

* **Quick and Easy Setup** - Ideal for those who need a simple, low-cost solution.

* **Cost-Effective** - Fewer resources are required, making it an attractive option for smaller operations.

#### Cons:

> * **High Risk** - With all control in the hands of one guardian, the risks of loss or theft are much higher. Additionally, if the server goes offline, users lose access to the federation until it’s restored.

Author: I realize that in the Cashu section under the mint I don't have a pros and cons list like there is for each Fedimint setup. I will go back and include a pros and cons list under the Cashu mint illustration and make sure to include this as well.

## When to Use Fedimint

> * **Federated Trust Models** - Use Fedimint when your application benefits from spreading trust across multiple entities.

Author: "I don’t mean to nitpick, but I’m unsure about using the word ‘mature’ here. It might unintentionally imply that Cashu is immature. Otherwise, I think the rest is solid. What do you think?

"Use Fedimint whenever a, custodial Lightning and ecash wallet is required where single points of failure aren't acceptable, i.e. when larger amounts are being stored or reliability is a concern."

> * **Federated Trust Models** - Use Fedimint when your application benefits from spreading trust across multiple entities.

Reviewer: > "Use Fedimint whenever a, custodial Lightning and ecash wallet is required where single points of failure aren't acceptable, i.e. when larger amounts are being stored or reliability is a concern."

Sounds good to me :)

> I don’t mean to nitpick, but I’m unsure about using the word ‘mature’ here.

Fine by me, maturity is a spectrum.

Cashu is more of a protocol spec after all (and that part seems solid), while Fedimint has one reference implementation by design, so comparisons are hard (I can't know all implementations). What I did hear during multiple conversations with Cashu developers at hackdays and such is that client state management is still an open issue, something we solved over a year ago on the Fedimint side. To be fair, while Fedimint is more complex by its nature we also had a head-start (first commit Jan 2021).

[Calle even tweetet about it recently](https://x.com/callebtc/status/1825907599486652924). So I don't think it's an unfair characterization from my side, but not worth fighting over.

**Products that use Fedimint**
- [Fedi](https://www.fedi.xyz)

> - [Clovyr](https://clovyr.app/)

Reviewer 2: Question (ducks): Should Mutinywallet be there if it's still shutting down by end of year?

> - [Clovyr](https://clovyr.app/)

Author: Thanks! I'll make sure to include in next commit.

> - [Clovyr](https://clovyr.app/)

Reviewer: I mean, the open source code will continue to exist and be usable independently afaik?

> - [Clovyr](https://clovyr.app/)

Author: Thanks! Just added this. Will include in next commit.

---
layout: guide
title: Fedimint
description: Learn the basics of Fedimint.
nav_order: 3
parent: Ecash
grand_parent: How it works
permalink: /guide/how-it-works/ecash/fedimint/
redirect_from:
 - /guide/ecash/fedimint
image: https://bitcoin.design/assets/images/guide/how-it-works/ecash/fedimint.jpg
main_classes: -no-top-padding
---

<!--

Editor's notes

Explains generally how fedimint works

Illustration sources

-->

{% include picture.html
   image = "/assets/images/guide/how-it-works/ecash/fedimint.jpg"
   retina = "/assets/images/guide/how-it-works/ecash/fedimint@2x.jpg"
   mobile = "/assets/images/guide/how-it-works/ecash/fedimint-mobile.jpg"
   mobileRetina = "/assets/images/guide/how-it-works/ecash/fedimint-mobile@2x.jpg"
   alt-text = "ecash overview illustration"
   width = 1600
   height = 700
   layout = "full-width"
%}

# Fedimint

> Fedimint is the first bitcoin-backed ecash protocol. Fedimint decentralizes trust across a federation of guardians, ensuring that no single guardian has complete control over user funds. Fedimint allows for various Federation configurations, each catering to different needs. These configurations include **Federated Mints**, **Federated Single Guardian Mints**, and **Solo Mints (Unfederated Mints)**, each with its strengths and trade-offs.

Author: Yes, Solomints are something I recently learned that Fedimint supports. They were created as an alternative to Cashu mints. In the first review call with the Fedimint developers, they referred to it as 'using a hammer to kill a fly,' since Fedimint is primarily designed for federated mints. However, in theory, you could use it for a solo mint if you wanted to, as the protocol does support it.

## Role of Guardians
Guardians are participants who collaboratively manage a federations operations. They secure the federation's funds using multi-sig wallets, validate transactions through consensus, and control ecash issuance and redemption.

> ## Role of Gateways

Reviewer 3: On the intro page, I mentioned that the illustration about connectivity using lighting felt out of place. It seems to fit well with this part about gateways. Isn't the gateway stuff generally a thing and maybe should not be in the fedimint page?

Author: ‘Gateway’ is a term used only for Fedimint. It’s unique because it’s a separate service that a federation needs to set up in order to be lightning compatible. Cashu doesn’t use ‘gateway.’  To connect a lightning node on Cashu the user simply edits a mint setup file and plugs in some mint info, like an admin macaroon and TLS cert (at least for LND—other implementations may require different details). Since gateways aren’t relevant for Cashu, I didn’t include them in the introduction.

I agree that this illustration is great for visualizing the role of the gateway. I’ll modify it to show only Fedimint federations so it’s clearer that it’s a Fedimint-specific service enabling lightning and connecting federations to the lightning network.

**How Gateways Work**: The gateway accepts bitcoin payments on the Lightning Network and converts them into the bitcoin-backed ecash tokens used within the federation. It can also convert bitcoin-backed ecash tokens into bitcoin and send them over the Lightning Network. This is crucial because it allows the users within a federation to interact with the outside bitcoin and Lightning Network.

**Vetted Gateways**: A vetted gateway is one that has been approved by the federation's guardians as reliable and trustworthy. This vetting process helps to ensure that users' transactions are handled by gateways that have a track record of reliability, reducing the likelihood of payment failures.

## Design Best Practices
To learn more about Fedimint specific design best practices you can check out the [Design best practices]({{ '/guide/how-it-works/ecash/design-best-practices' | relative_url }}) section.

## Types of Fedimints

### Federated Mints (4+ Guardians)

> {% include picture.html

Reviewer 3: I don't understand this diagram. There's a lot going on and I can't really tell what each actor does. I don't think the arrows are enough. One solution could be to have numbers in circles on there, along with a legend or text explanation that references them. Or, break it down into multiple diagrams.

Author: I agree. I’d like to revisit the illustrations. I think it would be clearer to depict where each mint or federation is its own circle, and these federations are part of a larger circle representing the Fedimint protocol. The protocol circle would then be connected to the entire bitcoin network with dotted lines representing lightning or on-chain. I’m drawing inspiration from these examples.

![image](https://github.com/user-attachments/assets/b0332db3-5cd6-47f7-a0cc-133e1ae02ef9)![image](https://github.com/user-attachments/assets/88db4734-7841-473d-b2fe-2509d246e8d7)

#### Pros:

* **Enhanced Resilience** - By using servers from multiple providers, the mint remains operational even during scheduled downtime or attempts of censorship, ensuring both reliability and security.

* **Simplified Setup** - Easier to set up compared to a fully federated mint, making it accessible while still offering improved resilience over a solo setup.

#### Cons:

* **Security Trade-off** - With only one guardian, there’s still a single point of failure, making it less secure than a fully federated setup.

### Solo Mints (Unfederated Mints)

{% include picture.html
   image = "/assets/images/guide/how-it-works/ecash/solomint@1x.jpg"
   retina = "/assets/images/guide/how-it-works/ecash/solomint@2x.jpg"
   mobile = "/assets/images/guide/how-it-works/ecash/solomint-mobile@2x.jpg"
   mobileRetina = "assets/images/guide/how-it-works/ecash/solomint-mobile@2x.jpg"
   alt-text = "An illustration of how users, ecash, and a solo mint interact."
   caption = "How a guardian interacts with users of a solo mint."
   width = 800
   height = 456
   layout = "full-width"
%}

> Solo mints are the simplest and most cost-effective option. Here, a single guardian manages the entire mint, making it easy to set up but with significant trade-offs in terms of security and resilience.

Reviewer 3: ```suggestion
Solo mints are the simplest and most cost-effective option. Here, a single guardian manages the entire mint, making it easy to set up, but with significant trade-offs in terms of security and resilience.
```

#### Pros:

* **Quick and Easy Setup** - Ideal for those who need a simple, low-cost solution.

* **Cost-Effective** - Fewer resources are required, making it an attractive option for smaller operations.

#### Cons:

* **High Risk** - With all control in the hands of one guardian, the risks of loss or theft are much higher. Additionally, if the server goes offline, users lose access to the federation until it’s restored.

## When to Use Fedimint

> * **Federated Trust Models** - Use Fedimint whenever a, custodial Lightning and ecash wallet is required where single points of failure aren't acceptable, i.e. when larger amounts are being stored or reliability is a concern.

Reviewer 3: ```suggestion
Use Fedimint whenever a custodial Lightning and ecash wallet is required, and single points of failure aren't acceptable. For example, when larger amounts are being stored or reliability is a concern.
```

No need for bullets when you just have a single point to make.

---
layout: guide
title: Fedimint
description: Learn the basics of Fedimint.
nav_order: 3
parent: Ecash
grand_parent: How it works
permalink: /guide/how-it-works/ecash/fedimint/
redirect_from:
 - /guide/ecash/fedimint
image: https://bitcoin.design/assets/images/guide/how-it-works/ecash/fedimint.jpg
main_classes: -no-top-padding
---

<!--

Editor's notes

Explains generally how fedimint works

Illustration sources

-->

{% include picture.html
   image = "/assets/images/guide/how-it-works/ecash/fedimint.jpg"
   retina = "/assets/images/guide/how-it-works/ecash/fedimint@2x.jpg"
   mobile = "/assets/images/guide/how-it-works/ecash/fedimint-mobile.jpg"
   mobileRetina = "/assets/images/guide/how-it-works/ecash/fedimint-mobile@2x.jpg"
   alt-text = "ecash overview illustration"
   width = 1600
   height = 600
   layout = "full-width"
%}

# Fedimint

> Fedimint is the first bitcoin-backed ecash protocol. Fedimint decentralizes trust across a federation of guardians, ensuring that no single guardian has complete control over user funds. Fedimint allows for various Federation configurations, each catering to different needs. These configurations include **Federated mints**, **Federated single guardian mints**, and **Solo mints (Unfederated mints)**, each with its strengths and trade-offs.

Reviewer 4: Notes from offline discussion: The distinction between federated, federated-single-guardian, and solo mints is an implementation detail. The user should make sure they have vetted the federation, should be able to view important metrics such as historical uptime, total lifespan, and recommendations/reviews from other users. Details such as the number of guardians are easily gamed.

## Role of guardians

> Guardians are participants who collaboratively manage a federations operations. They secure the federation's funds using multi-sig wallets, validate payments through consensus, and control ecash issuance and redemption.

Reviewer 4: Nit: "federation's operations"

## Role of gateways

> {% include picture.html

Reviewer 4: Notes from offline discussion: The graphic here could convey so much more information. I'd include the following:

* A gateway that serves multiple federations
* A gateway that serves a single federation
* A gateway with a lightning channel directly connected to another gateway
* A federation that is served by multiple gateways

We should also differentiate between real lightning channels and gateway<->federation connections

   image = "/assets/images/guide/how-it-works/ecash/gateways.png"
   retina = "/assets/images/guide/how-it-works/ecash/gateways@2x.png"
   mobile = "/assets/images/guide/how-it-works/ecash/gateways.png"
   mobileRetina = "/assets/images/guide/how-it-works/ecash/gateways@2x.png"
   alt-text = "Network diagram showing how gateways connect federations (represented by blue dots) with the lightning network (represented by orange dots). The nodes are linked by dashed lines, visualizing the interconnected structure of federations and lightning nodes."
   caption = "Gateways connect a federation’s ecash and bitcoin via the lightning network."
   width = 800
   height = 508
   layout = "full-width"
%}

> A gateway is a service that facilitates interactions between the federation (which operates largely off-chain) and the lightning network. A gateway acts as a bridge, enabling users within the federation to make payments to and receive payments on the lightning network.

Reviewer 4: I would get rid of "which operates largely off-chain". Maybe instead you could add a paragraph about why gateways are necessary, describing that without them, users could only operate entirely off-chain when transacting with other users within the same federation (through direct ecash swaps).

> **How gateways work**: The gateway accepts bitcoin payments on the lightning Network and converts them into the bitcoin-backed ecash tokens used within the federation. It can also convert bitcoin-backed ecash tokens into bitcoin and send them over the lightning network. This is crucial because it allows the users within a federation to interact with the outside bitcoin and lightning network.

Reviewer 4: Nit: "lightning Network" -> "lightning network"

> **Vetted gateways**: A vetted gateway is one that has been approved by the federation's guardians as reliable and trustworthy. This vetting process helps to ensure that users' payments are handled by gateways that have a track record of reliability, reducing the likelihood of payment failures.

Reviewer 4: The concept of unvetted gateways is being removed in the LNv2 module. All gateways will be vetted, since although they can be trusted to not steal user funds, they still must be trusted to provide a good user experience by routing quickly and not maliciously locking up funds in-flight. LNv2 won't be used until fedimint v0.6 (at least 3 months out) so it might be fine to just remove this - use your own discretion.

---
layout: guide
title: Fedimint
description: Learn the basics of Fedimint.
nav_order: 3
parent: Ecash
grand_parent: How it works
permalink: /guide/how-it-works/ecash/fedimint/
redirect_from:
 - /guide/ecash/fedimint
image: https://bitcoin.design/assets/images/guide/how-it-works/ecash/fedimint.jpg
main_classes: -no-top-padding
---

<!--

Editor's notes

Explains generally how fedimint works

Illustration sources

-->

{% include picture.html
   image = "/assets/images/guide/how-it-works/ecash/fedimint.jpg"
   retina = "/assets/images/guide/how-it-works/ecash/fedimint@2x.jpg"
   mobile = "/assets/images/guide/how-it-works/ecash/fedimint-mobile.jpg"
   mobileRetina = "/assets/images/guide/how-it-works/ecash/fedimint-mobile@2x.jpg"
   alt-text = "ecash overview illustration"
   width = 1600
   height = 600
   layout = "full-width"
%}

> # Fedimint

Reviewer 5: Sometimes Fedimint is capitalized and sometimes not.

Author: ack

Fedimint is the first bitcoin-backed ecash protocol. Fedimint decentralizes trust across a federation of guardians, ensuring that no single guardian has complete control over user funds. Fedimint allows for various Federation configurations, each catering to a community's specific needs and trust relationships

> ## Federated mints

Reviewer 5: there are three diagrams directly after each other, each discussing a different topic. It could be helpful to have short explainers for each of them, so that readers don't need to rely entirely on their own interpretation. It could also help to visually separate and structure the content.

   modalImage = "/assets/images/guide/how-it-works/ecash/federated-mint-3@2x.jpg"
   alt-text = "Diagram showing Federation 1 controlled by four guardians, each with a server and a set of keys. Federation 1 includes a user, an ecash icon, and a bitcoin icon connected by dashed lines."
   width = 800
   height = 500
   layout = "full-width"
%}

Federated Mints are the standard when it comes to security and resilience in the Fedimint protocol. Managed by four or more guardians, each running their own Fedimint server, this setup distributes control and minimizes risk.

#### Pros:

* **Security** - No single guardian can control the mint, reducing the risk of theft or mismanagement.

* **High resilience** - The mint remains operational even if one or more guardians are temporarily offline, ensuring continuous access.

#### Cons:

* **Complex setup** - Coordinating multiple guardians adds steps and cost.

## Role of guardians
Guardians are participants who collaboratively manage a federation's operations. They secure the federation's funds using multi-sig wallets, validate payments through consensus, and control ecash issuance and redemption.

## Role of gateways

> {% include picture.html

Reviewer 5: The usage of photos make it a bit harder to parse the diagram. Using the Bitcoin symbol to represent Fedimint ecash feels counter-intuitive.

   caption = "A Fedimint Gateway lets users pay lightning invoices with Fedimint bitcoin or receive Fedimint bitcoin from lightning payments, without converting between the two assets."
   width = 800
   height = 508
   layout = "full-width"
%}

A [gateway](https://fedimint.org/docs/GettingStarted/TechCompontents#lightning-swaps) is a service that facilitates interactions between the federation and the lightning network. A gateway acts as a bridge, enabling users within the federation to make payments to and receive payments on the lightning network.

**How gateways work**: The gateway accepts bitcoin payments on the lightning network and converts them into the bitcoin-backed ecash tokens used within the federation. It can also convert bitcoin-backed ecash tokens into bitcoin and send them over the lightning network. This is crucial because it allows the users within a federation to interact with the outside bitcoin and lightning network.

> 

Reviewer 5: Maybe move to the end of the page?

Author: My thinking is that since gateways are a core part of a federation, just like Guardians, they should be explained in a similar section. It makes sense to first explain what a federation is, then how it’s controlled, and finally how federations transfer money between each other (gateways). Gateways are already at the bottom of the explainer, and moving them even further down would feel out of place, as they wouldn’t be explained in the context of a federation anymore.

### Solo mints (unfederated mints)

{% include picture.html
   image = "/assets/images/guide/how-it-works/ecash/solomint.jpg"
   retina = "/assets/images/guide/how-it-works/ecash/solomint@2x.jpg"
   mobile = "/assets/images/guide/how-it-works/ecash/solomint-mobile.jpg"
   mobileRetina = "/assets/images/guide/how-it-works/ecash/solomint-mobile.jpg"
   alt-text = "An illustration of how users, ecash, and a solo mint interact."
   caption = "How a guardian interacts with users of a solo mint."
   width = 800
   height = 500
   layout = "full-width"
%}

Solo mints are the simplest and most cost-effective option. Here, a single guardian manages the entire mint, making it easy to set up, but with significant trade-offs in terms of security and resilience.

#### Pros:

* **Quick and easy setup** - Ideal for those who need a simple, low-cost solution.

* **Cost-effective** - Fewer resources are required, making it an attractive option for smaller operations.

#### Cons:

* **High risk** - With all control in the hands of one guardian, the risks of loss or theft are much higher. Additionally, if the server goes offline, users lose access to the federation until it’s restored. -->

> ## When to use fedimint

Reviewer 5: In other places, Fedimint is capitalized.
